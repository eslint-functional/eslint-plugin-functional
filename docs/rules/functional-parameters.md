# Enforce Functional Parameters (functional-parameters)

Disallow use of rest parameters, the `arguments` keyword and enforces that functions take a least 1 parameter.

## Rule Details

In functions, `arguments` is a special variable that is implicitly available.
This variable is an array containing the arguments passed to the function call and is often used to allow for any number of parameters to be passed to a function. Rest parameters are another way any number of parameters can be passed to a function.

When it comes to functional programming, known and explicit parameters must be used.

Note: With an unknown number of parameters, currying functions is a lot more difficult/impossible.

Examples of **incorrect** code for this rule:

```js
/* eslint functional/functional-parameters: "error" */

function add() {
  return arguments.reduce((sum, number) => sum + number, 0);
}
```

```js
/* eslint functional/functional-parameters: "error" */

function add(...numbers) {
  return numbers.reduce((sum, number) => sum + number, 0);
}
```

Examples of **correct** code for this rule:

```js
/* eslint functional/functional-parameters: "error" */

function add(numbers) {
  return numbers.reduce((sum, number) => sum + number, 0);
}
```

## Options

This rule accepts an options object of the following type:

```ts
{
  allowRestParameter: boolean;
  allowArgumentsKeyword: boolean;
  enforceParameterCount: false | "atLeastOne" | "exactlyOne" | {
    count: "atLeastOne" | "exactlyOne";
    ignoreIIFE: boolean;
  };
  ignorePattern?: string | Array<string>;
}
```

The default options:

```ts
{
  allowRestParameter: false,
  allowArgumentsKeyword: false,
  enforceParameterCount: {
    count: "atLeastOne",
    ignoreIIFE: true
  }
}
```

Note: the `lite` ruleset overrides the default options to:

```ts
{
  allowRestParameter: false,
  allowArgumentsKeyword: false,
  enforceParameterCount: false
}
```

### `allowRestParameter`

If true, this option allows for the use of rest parameters.

### `allowArgumentsKeyword`

If true, this option allows for the use of the `arguments` keyword.

### `enforceParameterCount`

This option enforces the number of parameters each function takes.

#### false

Don't enforce a parameter count.

#### "atLeastOne"

Require all functions to have a least one parameter.

There's not much point of having a function that doesn't take any parameters in functional programming.

#### "exactlyOne"

Require all functions to have exactly one parameter.

Any function that take takes multiple parameter can be rewritten as a higher-order function that only takes one.

Example:

```js
// This function
function add(x, y) {
  return x + y;
}

// can be rewritten like this.
function add(x) {
  return (y) => x + y;
}
```

See [Currying](https://en.wikipedia.org/wiki/Currying) and [Higher-order function](https://en.wikipedia.org/wiki/Higher-order_function) on Wikipedia for more infomation.

#### `enforceParameterCount.count`

See [enforceParameterCount](#enforceparametercount).

#### `enforceParameterCount.ignoreIIFE`

If true, this option allows for the use of [IIFEs](https://developer.mozilla.org/en-US/docs/Glossary/IIFE) that do not have any parameters.

### `ignorePattern`

Patterns will be matched against function names.
See the [ignorePattern](./options/ignore-pattern.md) docs for more infomation.
