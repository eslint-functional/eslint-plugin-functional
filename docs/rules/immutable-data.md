# Disallow mutating objects and arrays (immutable-data)

This rule prohibits syntax that mutates existing objects and arrays via assignment to or deletion of their properties/elements.

## Rule Details

While requiring the `readonly` modifier forces declared types to be immutable,
it won't stop assignment into or modification of untyped objects or external types declared under different rules.

Examples of **incorrect** code for this rule:

```js
/* eslint functional/immutable-data: "error" */

const obj = { foo: 1 };

obj.foo += 2; // <- Modifying an existing object/array is not allowed.
obj.bar = 1; // <- Modifying an existing object/array is not allowed.
delete obj.foo; // <- Modifying an existing object/array is not allowed.
Object.assign(obj, { bar: 2 }); // <- Modifying properties of existing object not allowed.
```

```js
/* eslint functional/immutable-data: "error" */

const arr = [0, 1, 2];

arr[0] = 4; // <- Modifying an array is not allowed.
arr.length = 1; // <- Modifying an array is not allowed.
arr.push(3); // <- Modifying an array is not allowed.
```

Examples of **correct** code for this rule:

```js
/* eslint functional/immutable-data: "error" */

const obj = { foo: 1 };
const arr = [0, 1, 2];

const x = {
  ...obj
  bar: [
    ...arr, 3, 4
  ]
}
```

## Options

This rule accepts an options object of the following type:

```ts
{
  assumeTypes:
    | boolean
    | {
        forArrays: boolean;
        forObjects: boolean;
      }
  ignoreImmediateMutation: boolean;
  ignorePattern?: string | Array<string>;
  ignoreAccessorPattern?: string | Array<string>;
}
```

The default options:

```ts
{
  assumeTypes: true,
  ignoreImmediateMutation: true,
};
```

### `assumeTypes`

The rule take advantage of TypeScript's typing engine to check if mutation is taking place.
If you are not using TypeScript, type checking cannot be performed; hence this option exists.

This option will make the rule assume the type of the nodes it is checking are of type Array/Object.  
However this may result in some false positives being picked up.

Disabling this option can result in false negatives, for example:

```js
// When this option is DISABLED (and type info is not available).
const x = [0, 1, 2];
x.push(3); // This will NOT be flagged.
// This is due to the fact that without a typing engine, we cannot tell that x is an array.
```

Note: This option will have no effect if the TypeScript typing engine is avaliable (i.e. you are using TypeScript and have configured eslint correctly).

### `ignoreImmediateMutation`

If true, immediate mutation of objects before they are assigned to a variable is allowed.
This allows for the use of array mutator methods to be chained to newly created arrays.

For example, an array can be immutably sorted like so:

```js
const original = ["foo", "bar", "baz"];
const sorted = [...original].sort((a, b) => a.localeCompare(b)); // This is OK with ignoreImmediateMutation.
```

### `ignorePattern`

Patterns will be matched against variable names.
See the [ignorePattern](./options/ignore-pattern.md) docs for more infomation.

### `ignoreAccessorPattern`

This option takes a match string or an array of match strings (not a RegExp pattern).

The match string allows you to specify dot seperated `.` object paths and has support for "glob" `*` and "globstar" `**` matching.

For example:

```js
{
  // Ignore all reassigning to object properties that are prefixed with "mutable_".
  "ignoreAccessorPattern": "**.mutable_*"
}
```

```js
{
  // Ignore all shallow mutations made to object properties that are prefixed with "mutable_".
  "ignoreAccessorPattern": "**.mutable_*.*"
}
```

```js
{
  // Ignore all deep mutations made to object properties that are prefixed with "mutable_".
  "ignoreAccessorPattern": "**.mutable_*.*.**"
}
```

```js
{
  // Ignore all deep mutations and reassigning to object properties that are prefixed with "mutable_".
  "ignoreAccessorPattern": "**.mutable_*.**"
  // This is the same as `"ignoreAccessorPattern": ["**.mutable_*", "**.mutable_*.*.**"]`
}
```

#### Wildcards

The following wildcards can be used when specifing a pattern:

`**` - Match any depth (including zero). Can only be used as a full accessor.
`*` - When used as a full accessor, match the next accessor (there must be one). When used as part of an accessor, match any characters.
