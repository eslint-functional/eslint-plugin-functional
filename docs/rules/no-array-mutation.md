### no-array-mutation

[![Type Info Required][type-info-badge]][type-info-url]

This rule prohibits mutating an array via assignment to or deletion of their elements/properties. This rule enforces array immutability without the use of `ReadonlyArray<T>` (as apposed to [readonly-array](#readonly-array)).

```typescript
const x = [0, 1, 2];

x[0] = 4; // <- Mutating an array is not allowed.
x.length = 1; // <- Mutating an array is not allowed.
x.push(3); // <- Mutating an array is not allowed.
```

#### Has Fixer

No

#### Options

- [ignore-prefix](#using-the-ignore-prefix-option)
- [ignore-suffix](#using-the-ignore-suffix-option)
- [ignore-pattern](#using-the-ignore-pattern-option)
- [ignore-new-array](#using-the-ignore-new-array-option-with-no-array-mutation)
- ~~ignore-mutation-following-accessor~~ - _deprecated in favor of [ignore-new-array](#using-the-ignore-new-array-option-with-no-array-mutation)_

#### Example config

```javascript
"no-array-mutation": true
```

```javascript
"no-array-mutation": [true, {"ignore-prefix": "mutable"}]
```

```javascript
"no-array-mutation": [true, "ignore-new-array"]
```
