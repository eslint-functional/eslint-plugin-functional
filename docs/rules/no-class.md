# Disallow classes (no-class)

Disallow use of the `class` keyword.

## Rule Details

Thanks to libraries like [recompose](https://github.com/acdlite/recompose) and Redux's [React Container components](http://redux.js.org/docs/basics/UsageWithReact.html), there's not much reason to build Components using `React.createClass` or ES6 classes anymore. The `no-this` rule makes this explicit.

```typescript
const Message = React.createClass({
  render: function() {
    return <div>{this.props.message}</div>; // <- no this allowed
  }
});
```

Instead of creating classes, you should use React 0.14's [Stateless Functional Components](https://medium.com/@joshblack/stateless-components-in-react-0-14-f9798f8b992d#.t5z2fdit6) and save yourself some keystrokes:

```typescript
const Message = ({ message }) => <div>{message}</div>;
```

What about lifecycle methods like `shouldComponentUpdate`? We can use the [recompose](https://github.com/acdlite/recompose) library to apply these optimizations to your Stateless Functional Components. The [recompose](https://github.com/acdlite/recompose) library relies on the fact that your Redux state is immutable to efficiently implement shouldComponentUpdate for you.

```typescript
import { pure, onlyUpdateForKeys } from "recompose";

const Message = ({ message }) => <div>{message}</div>;

// Optimized version of same component, using shallow comparison of props
// Same effect as React's PureRenderMixin
const OptimizedMessage = pure(Message);

// Even more optimized: only updates if specific prop keys have changed
const HyperOptimizedMessage = onlyUpdateForKeys(["message"], Message);
```

## Options

The rule does not accept any options.
