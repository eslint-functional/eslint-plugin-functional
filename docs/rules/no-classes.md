<!-- markdownlint-disable -->
<!-- begin auto-generated rule header -->

# Disallow classes (`functional/no-classes`)

üíº This rule is enabled in the following configs: ‚òëÔ∏è `lite`, `noOtherParadigms`, ‚úÖ `recommended`, üîí `strict`.

<!-- end auto-generated rule header -->
<!-- markdownlint-restore -->
<!-- markdownlint-restore -->

Disallow use of the `class` keyword.

## Rule Details

### ‚ùå Incorrect

<!-- eslint-skip -->

```js
/* eslint functional/no-classes: "error" */

class Dog {
  constructor(name, age) {
    this.name = name;
    this.age = age;
  }

  get ageInDogYears() {
    return 7 * this.age;
  }
}

const dogA = new Dog("Jasper", 2);

console.log(`${dogA.name} is ${dogA.ageInDogYears} in dog years.`);
```

### ‚úÖ Correct

```js
/* eslint functional/no-classes: "error" */

function getAgeInDogYears(age) {
  return 7 * age;
}

const dogA = {
  name: "Jasper",
  age: 2,
};

console.log(`${dogA.name} is ${getAgeInDogYears(dogA.age)} in dog years.`);
```

### React Examples

Thanks to libraries like [recompose](https://github.com/acdlite/recompose) and Redux's
[React Container components](http://redux.js.org/docs/basics/UsageWithReact.html), there's not much reason to build
Components using `React.createClass` or ES6 classes anymore. The `no-this-expressions` rule makes this explicit.

```js
const Message = React.createClass({
  render() {
    return <div>{this.props.message}</div>; // <- no this allowed
  },
});
```

Instead of creating classes, you should use React 0.14's
[Stateless Functional
Components](https://medium.com/@joshblack/stateless-components-in-react-0-14-f9798f8b992d#.t5z2fdit6)
and save yourself some keystrokes:

```js
const Message = ({ message }) => <div>{message}</div>;
```

What about lifecycle methods like `shouldComponentUpdate`?
We can use the [recompose](https://github.com/acdlite/recompose) library to apply these optimizations to your
Stateless Functional Components. The [recompose](https://github.com/acdlite/recompose) library relies on the fact that
your Redux state is immutable to efficiently implement `shouldComponentUpdate` for you.

```js
import { onlyUpdateForKeys, pure } from "recompose";

const Message = ({ message }) => <div>{message}</div>;

// Optimized version of same component, using shallow comparison of props
// Same effect as React's PureRenderMixin
const OptimizedMessage = pure(Message);

// Even more optimized: only updates if specific prop keys have changed
const HyperOptimizedMessage = onlyUpdateForKeys(["message"], Message);
```
