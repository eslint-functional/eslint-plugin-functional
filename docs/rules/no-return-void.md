# Disallow Returning Nothing (no-return-void)

Disallow functions that are declared as returning nothing.

## Rule Details

In functional programming functions must return something, they cannot return nothing.

By default, this rule allows function to return `undefined` and `null`.

Note: For performance reasons, this rule does not check implicit return types.
We recommend using the rule [@typescript-eslint/explicit-function-return-type](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/explicit-function-return-type.md) in conjunction with this rule.

Examples of **incorrect** code for this rule:

```ts
/* eslint functional/no-return-void: "error" */

function updateText(): void {

}
```

Examples of **correct** code for this rule:

```ts
/* eslint functional/no-return-void: "error" */

function updateText(value: string): string {

}
```

## Options

This rule accepts an options object of the following type:

```ts
{
  allowNull: boolean;
  allowUndefined: boolean;
}
```

The default options:

```ts
{
  allowNull: true,
  allowUndefined: true
}
```

### allowNull

If true allow returning null.

### allowUndefined

If true allow returning undefined.
