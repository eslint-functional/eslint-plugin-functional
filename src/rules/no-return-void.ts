import type { TSESTree } from "@typescript-eslint/experimental-utils";
import type { JSONSchema4 } from "json-schema";

import type { RuleContext, RuleMetaData, RuleResult } from "~/utils/rule";
import { createRule } from "~/utils/rule";
import {
  isTSNullKeyword,
  isTSUndefinedKeyword,
  isTSVoidKeyword,
} from "~/utils/typeguard";

// The name of this rule.
export const name = "no-return-void" as const;

// The options this rule can take.
type Options = {
  readonly allowNull: boolean;
  readonly allowUndefined: boolean;
};

// The schema for the rule options.
const schema: JSONSchema4 = [
  {
    type: "object",
    properties: {
      allowNull: {
        type: "boolean",
      },
      allowUndefined: {
        type: "boolean",
      },
    },
    additionalProperties: false,
  },
];

// The default options for the rule.
const defaultOptions: Options = {
  allowNull: true,
  allowUndefined: true,
};

// The possible error messages.
const errorMessages = {
  generic: "Function must return a value.",
} as const;

// The meta data for this rule.
const meta: RuleMetaData<keyof typeof errorMessages> = {
  type: "suggestion",
  docs: {
    description: "Disallow functions that don't return anything.",
    category: "Best Practices",
    recommended: "error",
  },
  messages: errorMessages,
  schema,
};

/**
 * Check if the given function node violates this rule.
 */
function checkFunction(
  node:
    | TSESTree.FunctionDeclaration
    | TSESTree.FunctionExpression
    | TSESTree.ArrowFunctionExpression
    | TSESTree.TSFunctionType,
  context: RuleContext<keyof typeof errorMessages, Options>,
  options: Options
): RuleResult<keyof typeof errorMessages, Options> {
  return {
    context,
    descriptors:
      node.returnType !== undefined &&
      (isTSVoidKeyword(node.returnType.typeAnnotation) ||
        (!options.allowNull &&
          isTSNullKeyword(node.returnType.typeAnnotation)) ||
        (!options.allowUndefined &&
          isTSUndefinedKeyword(node.returnType.typeAnnotation)))
        ? [{ node: node.returnType, messageId: "generic" }]
        : [],
  };
}

// Create the rule.
export const rule = createRule<keyof typeof errorMessages, Options>(
  name,
  meta,
  defaultOptions,
  {
    FunctionDeclaration: checkFunction,
    FunctionExpression: checkFunction,
    ArrowFunctionExpression: checkFunction,
    TSFunctionType: checkFunction,
  }
);
