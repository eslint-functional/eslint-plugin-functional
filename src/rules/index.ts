import * as functionalParameters from "./functional-parameters";
import * as immutableData from "./immutable-data";
import * as noClass from "./no-class";
import * as noConditionalStatement from "./no-conditional-statement";
import * as noExpressionStatement from "./no-expression-statement";
import * as noLet from "./no-let";
import * as noLoop from "./no-loop-statement";
import * as noMethodSignature from "./no-method-signature";
import * as noMixedType from "./no-mixed-type";
import * as noPromiseReject from "./no-promise-reject";
import * as noReturnVoid from "./no-return-void";
import * as noThisExpression from "./no-this-expression";
import * as noThrowStatement from "./no-throw-statement";
import * as noTryStatement from "./no-try-statement";
import * as preferReadonlyTypes from "./prefer-readonly-type";
import * as preferType from "./prefer-type-literal";

/**
 * All of the custom rules.
 */
export const rules = {
  [functionalParameters.name]: functionalParameters.rule,
  [immutableData.name]: immutableData.rule,
  [noClass.name]: noClass.rule,
  [noConditionalStatement.name]: noConditionalStatement.rule,
  [noExpressionStatement.name]: noExpressionStatement.rule,
  [noLet.name]: noLet.rule,
  [noLoop.name]: noLoop.rule,
  [noMethodSignature.name]: noMethodSignature.rule,
  [noMixedType.name]: noMixedType.rule,
  [noPromiseReject.name]: noPromiseReject.rule,
  [noReturnVoid.name]: noReturnVoid.rule,
  [noThisExpression.name]: noThisExpression.rule,
  [noThrowStatement.name]: noThrowStatement.rule,
  [noTryStatement.name]: noTryStatement.rule,
  [preferReadonlyTypes.name]: preferReadonlyTypes.rule,
  [preferType.name]: preferType.rule,
};
