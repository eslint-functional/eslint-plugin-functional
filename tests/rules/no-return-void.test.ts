/**
 * @file Tests for no-return-void.
 */

import dedent from "dedent";
import { RuleTester } from "eslint";

import { name, rule } from "../../src/rules/no-return-void";

import { typescript } from "../helpers/configs";
import {
  InvalidTestCase,
  processInvalidTestCase,
  processValidTestCase,
  tsInstalled,
  ValidTestCase
} from "../helpers/util";

// Valid test cases.
const valid: ReadonlyArray<ValidTestCase> = [
  {
    code: dedent`
      function foo(bar) {
        return bar + 1;
      }`,
    optionsSet: [[], [{ allowNull: false }], [{ allowUndefined: false }]]
  },
  {
    code: dedent`
      function foo(bar: number): number {
        return bar + 1;
      }`,
    optionsSet: [[], [{ allowNull: false }], [{ allowUndefined: false }]]
  },
  // Not testing implicit return types, therefore this is valid.
  {
    code: dedent`
      function foo(bar) {
        console.log(bar);
      }`,
    optionsSet: [[], [{ allowNull: false }], [{ allowUndefined: false }]]
  },
  // Allow null.
  {
    code: dedent`
      function foo(): null {
        return null;
      }`,
    optionsSet: [[], [{ allowNull: true }], [{ allowUndefined: false }]]
  },
  // Allow undefined.
  {
    code: dedent`
      function foo(): undefined {
        return undefined;
      }`,
    optionsSet: [[], [{ allowNull: false }], [{ allowUndefined: true }]]
  }
];

// Invalid test cases.
const invalid: ReadonlyArray<InvalidTestCase> = [
  // Disallow void.
  {
    code: dedent`
      function foo(bar: number): void {
        console.log(bar);
      }`,
    optionsSet: [[]],
    errors: [
      {
        messageId: "generic",
        type: "TSTypeAnnotation",
        line: 1,
        column: 26
      }
    ]
  },
  // Disallow undefined.
  {
    code: dedent`
      function foo(bar: number): undefined {
        console.log(bar);
        return undefined;
      }`,
    optionsSet: [[{ allowUndefined: false }]],
    errors: [
      {
        messageId: "generic",
        type: "TSTypeAnnotation",
        line: 1,
        column: 26
      }
    ]
  },
  // Disallow null.
  {
    code: dedent`
      function foo(bar: number): null {
        console.log(bar);
        return null;
      }`,
    optionsSet: [[{ allowNull: false }]],
    errors: [
      {
        messageId: "generic",
        type: "TSTypeAnnotation",
        line: 1,
        column: 26
      }
    ]
  },
  // Disallow higher-order function void.
  {
    code: dedent`
      function foo(bar: number): (baz: number) => void {
        return baz => { console.log(bar, baz); }
      }`,
    optionsSet: [[]],
    errors: [
      {
        messageId: "generic",
        type: "TSTypeAnnotation",
        line: 1,
        column: 42
      }
    ]
  }
];

if (tsInstalled()) {
  describe("TypeScript", () => {
    const ruleTester = new RuleTester(typescript);
    ruleTester.run(name, rule, {
      valid: processValidTestCase(valid),
      invalid: processInvalidTestCase(invalid)
    });
  });
}
