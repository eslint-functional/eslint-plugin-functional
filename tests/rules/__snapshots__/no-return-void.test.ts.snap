// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`no-return-void > typescript > options > allowNull > reports null returning functions when disallowed 1`] = `
{
  "fixed": false,
  "messages": [
    {
      "column": 26,
      "endColumn": 32,
      "endLine": 1,
      "line": 1,
      "message": "Function must return a value.",
      "messageId": "generic",
      "nodeType": "TSTypeAnnotation",
      "ruleId": "no-return-void",
      "severity": 2,
    },
  ],
  "output": "function foo(bar: number): null {
  return null;
}",
  "steps": [
    {
      "fixed": false,
      "messages": [
        {
          "column": 26,
          "endColumn": 32,
          "endLine": 1,
          "line": 1,
          "message": "Function must return a value.",
          "messageId": "generic",
          "nodeType": "TSTypeAnnotation",
          "ruleId": "rule-to-test/no-return-void",
          "severity": 2,
        },
      ],
      "output": "function foo(bar: number): null {
  return null;
}",
    },
  ],
}
`;

exports[`no-return-void > typescript > options > allowNull > reports null returning functions with inferred return type when disallowed 1`] = `
{
  "fixed": false,
  "messages": [
    {
      "column": 1,
      "endColumn": 2,
      "endLine": 3,
      "line": 1,
      "message": "Function must return a value.",
      "messageId": "generic",
      "nodeType": "FunctionDeclaration",
      "ruleId": "no-return-void",
      "severity": 2,
    },
  ],
  "output": "function foo(bar: number) {
  return null;
}",
  "steps": [
    {
      "fixed": false,
      "messages": [
        {
          "column": 1,
          "endColumn": 2,
          "endLine": 3,
          "line": 1,
          "message": "Function must return a value.",
          "messageId": "generic",
          "nodeType": "FunctionDeclaration",
          "ruleId": "rule-to-test/no-return-void",
          "severity": 2,
        },
      ],
      "output": "function foo(bar: number) {
  return null;
}",
    },
  ],
}
`;

exports[`no-return-void > typescript > options > allowUndefined > reports undefined returning functions when disallowed 1`] = `
{
  "fixed": false,
  "messages": [
    {
      "column": 26,
      "endColumn": 37,
      "endLine": 1,
      "line": 1,
      "message": "Function must return a value.",
      "messageId": "generic",
      "nodeType": "TSTypeAnnotation",
      "ruleId": "no-return-void",
      "severity": 2,
    },
  ],
  "output": "function foo(bar: number): undefined {
  return undefined;
}",
  "steps": [
    {
      "fixed": false,
      "messages": [
        {
          "column": 26,
          "endColumn": 37,
          "endLine": 1,
          "line": 1,
          "message": "Function must return a value.",
          "messageId": "generic",
          "nodeType": "TSTypeAnnotation",
          "ruleId": "rule-to-test/no-return-void",
          "severity": 2,
        },
      ],
      "output": "function foo(bar: number): undefined {
  return undefined;
}",
    },
  ],
}
`;

exports[`no-return-void > typescript > options > allowUndefined > reports undefined returning functions with inferred return type when disallowed 1`] = `
{
  "fixed": false,
  "messages": [
    {
      "column": 1,
      "endColumn": 2,
      "endLine": 3,
      "line": 1,
      "message": "Function must return a value.",
      "messageId": "generic",
      "nodeType": "FunctionDeclaration",
      "ruleId": "no-return-void",
      "severity": 2,
    },
  ],
  "output": "function foo(bar: number) {
  return undefined;
}",
  "steps": [
    {
      "fixed": false,
      "messages": [
        {
          "column": 1,
          "endColumn": 2,
          "endLine": 3,
          "line": 1,
          "message": "Function must return a value.",
          "messageId": "generic",
          "nodeType": "FunctionDeclaration",
          "ruleId": "rule-to-test/no-return-void",
          "severity": 2,
        },
      ],
      "output": "function foo(bar: number) {
  return undefined;
}",
    },
  ],
}
`;

exports[`no-return-void > typescript > reports void returning functions 1`] = `
{
  "fixed": false,
  "messages": [
    {
      "column": 26,
      "endColumn": 32,
      "endLine": 1,
      "line": 1,
      "message": "Function must return a value.",
      "messageId": "generic",
      "nodeType": "TSTypeAnnotation",
      "ruleId": "no-return-void",
      "severity": 2,
    },
  ],
  "output": "function foo(bar: number): void {
  console.log(bar);
}",
  "steps": [
    {
      "fixed": false,
      "messages": [
        {
          "column": 26,
          "endColumn": 32,
          "endLine": 1,
          "line": 1,
          "message": "Function must return a value.",
          "messageId": "generic",
          "nodeType": "TSTypeAnnotation",
          "ruleId": "rule-to-test/no-return-void",
          "severity": 2,
        },
      ],
      "output": "function foo(bar: number): void {
  console.log(bar);
}",
    },
  ],
}
`;

exports[`no-return-void > typescript > reports void returning functions with inferred return type 1`] = `
{
  "fixed": false,
  "messages": [
    {
      "column": 1,
      "endColumn": 2,
      "endLine": 3,
      "line": 1,
      "message": "Function must return a value.",
      "messageId": "generic",
      "nodeType": "FunctionDeclaration",
      "ruleId": "no-return-void",
      "severity": 2,
    },
  ],
  "output": "function foo(bar: number) {
  console.log(bar);
}",
  "steps": [
    {
      "fixed": false,
      "messages": [
        {
          "column": 1,
          "endColumn": 2,
          "endLine": 3,
          "line": 1,
          "message": "Function must return a value.",
          "messageId": "generic",
          "nodeType": "FunctionDeclaration",
          "ruleId": "rule-to-test/no-return-void",
          "severity": 2,
        },
      ],
      "output": "function foo(bar: number) {
  console.log(bar);
}",
    },
  ],
}
`;
